packages/runtime/container-runtime/src/dataStores.ts

```diff
@@ -0,0 +1,243 @@
+                dataStoreContext = new RemotedFluidDataStoreContext(
+                    key,
+                    typeof value === "string" ? value : Promise.resolve(value),
+                    this.runtime,
+                    this.runtime.storage,
+                    this.runtime.scope,
+                    this.summaryTracker.createOrGetChild(key, this.summaryTracker.referenceSequenceNumber),
+                    this.getCreateChildSummarizerNodeFn(key, { type: CreateSummarizerNodeSource.FromSummary }));
+            } else {
+                let pkgFromSnapshot: string[];
+                if (typeof value !== "object") {
+                    throw new Error("Snapshot should be there to load from!!");
+                }
+                const snapshotTree = value;
+                // Need to rip through snapshot.
+                const { pkg, snapshotFormatVersion, isRootDataStore }
+                    = readAndParseFromBlobs<IFluidDataStoreAttributes>(
+                        snapshotTree.blobs,
+                        snapshotTree.blobs[".component"]);
+                // Use the snapshotFormatVersion to determine how the pkg is encoded in the snapshot.
+                // For snapshotFormatVersion = "0.1", pkg is jsonified, otherwise it is just a string.
+                // However the feature of loading a detached container from snapshot, is added when the
+                // snapshotFormatVersion is "0.1", so we don't expect it to be anything else.
+                if (snapshotFormatVersion === currentSnapshotFormatVersion) {
+                    pkgFromSnapshot = JSON.parse(pkg) as string[];
+                } else {
+                    throw new Error(`Invalid snapshot format version ${snapshotFormatVersion}`);
+                }
+
+                /**
+                 * If there is no isRootDataStore in the attributes blob, set it to true. This will ensure that data
+                 * stores in older documents are not garbage collected incorrectly. This may lead to additional roots
+                 * in the document but they won't break.
+                 */
+                dataStoreContext = new LocalFluidDataStoreContext(
+                    key,
+                    pkgFromSnapshot,
+                    this.runtime,
+                    this.runtime.storage,
+                    this.runtime.scope,
+                    this.summaryTracker.createOrGetChild(key, this.runtime.deltaManager.lastSequenceNumber),
+                    this.getCreateChildSummarizerNodeFn(key, { type: CreateSummarizerNodeSource.FromSummary }),
+                    (cr: IFluidDataStoreChannel) => this.bindFluidDataStore(cr),
+                    snapshotTree,
+                    isRootDataStore ?? true);
+            }
+            this.contexts.addBoundOrRemote(key, dataStoreContext);
+        }
+    }
+
+    public processAttachMessage(message: ISequencedDocumentMessage, local: boolean) {
+        const attachMessage = message.contents;
```

Comment:
Why get rid of the `as`? IMO doing a `.` on an `any` is just as unpleasant as casting via `as`. Should there be some type guard with a runtime check/error instead?

Reply:
not sure what happend,as i just copy pasted. will re-add

Reply:
🤖